sdks/universal-router-sdk/src/entities/actions/uniswap.ts

```diff
@@ -345,49 +348,80 @@ function addMixedSwap<TInput extends Currency, TOutput extends Currency>(
-      return route.pools.every((pool) => pool instanceof V3Pool)
+    if (isLastSectionInRoute(i)) {
+      nextInputToken = outputToken
+      swapRecipient = tradeRecipient
+    } else {
+      const nextPool = sections[i + 1][0]
+      nextInputToken = getPathCurrency(outputToken, nextPool)
+
+      const v2PoolIsSwapRecipient = nextPool instanceof Pair && outputToken.equals(nextInputToken)
+      swapRecipient = v2PoolIsSwapRecipient ? (nextPool as Pair).liquidityToken.address : ROUTER_AS_RECIPIENT
     }
 
-    if (mixedRouteIsAllV3(newRoute)) {
-      const path: string = encodeMixedRouteToPath(newRoute)
+    if (routePool instanceof V4Pool) {
+      const v4Planner = new V4Planner()
+      const v4SubRoute = new V4Route(section as V4Pool[], subRoute.input, subRoute.output)
+
+      v4Planner.addSettle(inputToken, payerIsUser && i === 0, (i == 0 ? amountIn : CONTRACT_BALANCE) as BigNumber)
+      v4Planner.addAction(Actions.SWAP_EXACT_IN, [
+        {
+          currencyIn: inputToken.isNative ? ETH_ADDRESS : inputToken.address,
+          path: encodeV4RouteToPath(v4SubRoute),
+          amountIn: 0, // denotes open delta, amount set in v4Planner.addSettle()
+          amountOutMinimum: !isLastSectionInRoute(i) ? 0 : amountOut,
+        },
+      ])
+      v4Planner.addTake(outputToken, swapRecipient)
 
+      planner.addCommand(CommandType.V4_SWAP, [v4Planner.finalize()])
+    } else if (routePool instanceof V3Pool) {
       planner.addCommand(CommandType.V3_SWAP_EXACT_IN, [
-        // if not last section: send tokens directly to the first v2 pair of the next section
-        // note: because of the partitioning function we can be sure that the next section is v2
-        isLastSectionInRoute(i) ? tradeRecipient : (sections[i + 1][0] as Pair).liquidityToken.address,
+        swapRecipient, // recipient
         i == 0 ? amountIn : CONTRACT_BALANCE, // amountIn
         !isLastSectionInRoute(i) ? 0 : amountOut, // amountOut
-        path, // path
+        encodeMixedRouteToPath(subRoute), // path
         payerIsUser && i === 0, // payerIsUser
       ])
-    } else {
+    } else if (routePool instanceof Pair) {
       planner.addCommand(CommandType.V2_SWAP_EXACT_IN, [
-        isLastSectionInRoute(i) ? tradeRecipient : ROUTER_AS_RECIPIENT, // recipient
+        swapRecipient, // recipient
         i === 0 ? amountIn : CONTRACT_BALANCE, // amountIn
         !isLastSectionInRoute(i) ? 0 : amountOut, // amountOutMin
-        newRoute.path.map((token) => token.wrapped.address), // path
+        subRoute.path.map((token) => token.wrapped.address), // path
         payerIsUser && i === 0,
       ])
+    } else {
+      throw new Error('Unexpected Pool Type')
+    }
+
+    // perform a token transition (wrap/unwrap if necessary)
+    if (!isLastSectionInRoute(i)) {
+      if (outputToken.isNative && !nextInputToken.isNative) {
+        planner.addCommand(CommandType.WRAP_ETH, [ROUTER_AS_RECIPIENT, CONTRACT_BALANCE])
```

Comment:
ok to add, since we won't support special ETH/WETH hook pool for the initial launch. On the SOR/routing-api level, it's even less flexible, that it won't produce a route from ETH -> WETH or WETH -> ETH in a pure v4 route, or a mixed route with pools that includes any ETH -> WETH or WETH -> ETH path.

I guess we can add a TODO linear ticket to remind ourselves, that once we support  special ETH/WETH hook pool, revisit the wrap/unwrap here to see if it's necessary.

Reply:
we sync'ed in slack. sorry for bringing up on v4 route part, that confuses discussions. My bcc to here - 

> SOR doesnt support mixed route with the ETH/WETH, where in the sdks, I'm aware you made the ETH/WETH route discoverable in the mixed route via this pr https://github.com/Uniswap/sdks/pull/81.

> sorry for bringing v4 route part, that only confuses the discussions. shouldve focus only on v4.
> v2/v3 --> v4 WETH unwraps to ETH
> v4 --> v2/v3 ETH wraps to WETH
> SOR won't return a mixed route like above, because of the SOR implementation complexity.

and we resolved the confusions.